File: backend/src/apps/iam/models/token_tracking.py
from typing import Optional, TYPE_CHECKING
from datetime import datetime
from sqlmodel import SQLModel, Field, Relationship
from sqlalchemy import Column, Enum as SAEnum
from src.apps.core.security import TokenType

if TYPE_CHECKING:
    from .user import User


class TokenTrackingBase(SQLModel):
    token_jti: str = Field(
        max_length=255,
        index=True,
        unique=True,
        description="JWT ID (JTI) - unique identifier for the token"
    )
    token_type: TokenType = Field(
        sa_column=Column(SAEnum(TokenType, values_callable=lambda e: [m.value for m in e])),
        description="Type of token: access, refresh"
    )
    ip_address: str = Field(
        max_length=45,
        description="IP address where token was issued"
    )
    user_agent: str = Field(
        max_length=255,
        description="User agent string when token was issued"
    )
    is_active: bool = Field(
        default=True,
        description="Whether the token is still active"
    )
    revoked_at: Optional[datetime] = Field(
        default=None,
        description="When the token was revoked"
    )
    revoke_reason: str = Field(
        default="",
        max_length=255,
        description="Reason for token revocation"
    )
    expires_at: datetime = Field(
        description="When the token expires"
    )
    created_at: datetime = Field(
        default_factory=datetime.now,
        description="When the token was created"
    )


class TokenTracking(TokenTrackingBase, table=True):
    """
    Model to track issued tokens for users, including their status and metadata.
    """
    id: int = Field(
        default=None,
        primary_key=True,
        description="Unique identifier for the token tracking entry"
    )
    user_id: Optional[int] = Field(
        foreign_key="user.id",
        ondelete="CASCADE",
        description="ID of the user this token belongs to"
    )

    # Relationships
    user: Optional[User] = Relationship(back_populates="tokens")

File: mobile/lib/features/profile/data/models/token_tracking.dart
enum TokenType {
  access,
  refresh,
  passwordReset,
  emailVerification,
  tempAuth,
  bearer;

  static TokenType fromString(String v) {
    switch (v) {
      case 'access':
        return TokenType.access;
      case 'refresh':
        return TokenType.refresh;
      case 'password_reset':
        return TokenType.passwordReset;
      case 'email_verification':
        return TokenType.emailVerification;
      case 'temp_auth':
        return TokenType.tempAuth;
      case 'bearer':
        return TokenType.bearer;
      default:
        return TokenType.access;
    }
  }

  String toJson() {
    switch (this) {
      case TokenType.access:
        return 'access';
      case TokenType.refresh:
        return 'refresh';
      case TokenType.passwordReset:
        return 'password_reset';
      case TokenType.emailVerification:
        return 'email_verification';
      case TokenType.tempAuth:
        return 'temp_auth';
      case TokenType.bearer:
        return 'bearer';
    }
  }
}

class TokenTracking {
  final String id;
  final String userId;
  final String tokenJti;
  final TokenType tokenType;
  final String ipAddress;
  final String userAgent;
  final bool isActive;
  final String? revokedAt;
  final String revokeReason;
  final String expiresAt;
  final String createdAt;

  const TokenTracking({
    required this.id,
    required this.userId,
    required this.tokenJti,
    required this.tokenType,
    required this.ipAddress,
    required this.userAgent,
    required this.isActive,
    this.revokedAt,
    required this.revokeReason,
    required this.expiresAt,
    required this.createdAt,
  });

  factory TokenTracking.fromJson(Map<String, dynamic> json) {
    return TokenTracking(
      id: json['id'].toString(),
      userId: json['user_id']?.toString() ?? '',
      tokenJti: json['token_jti'] as String? ?? '',
      tokenType: TokenType.fromString(json['token_type'] as String? ?? 'access'),
      ipAddress: json['ip_address'] as String? ?? '',
      userAgent: json['user_agent'] as String? ?? '',
      isActive: json['is_active'] as bool? ?? true,
      revokedAt: json['revoked_at'] as String?,
      revokeReason: json['revoke_reason'] as String? ?? '',
      expiresAt: json['expires_at'] as String? ?? '',
      createdAt: json['created_at'] as String? ?? '',
    );
  }
}
